backend: Add failed scheduled message notification and update API dict for `failed` boolean.

laurynmm opened the pull request:
Adds sending a message from Notification Bot to the user who scheduled the message if/when there is an error sending the message. The only errors that do not send these notification messages are when the realm or user sending the message has been deactivated.

Also, updates the API scheduled message dicts to have a `failed` field that will be `True` in the case that the server tried to send the scheduled message but there was an error. The same commit fixes two colons in the feature level 179 API changelog entry as well.

This is part of work on #25501. After these changes, the remaining task in that issue is to update the scheduled messages overlay and UI for the new icon.

---

**Screenshots and screen captures:**

<details>
<summary>Example of Notification Bot message</summary>

#### The yellow highlight is from my search filter as I had some other test messages, but wanted the bot avatar in the screenshot

![Screenshot from 2023-05-11 20-59-21](https://github.com/zulip/zulip/assets/63245456/db988709-2180-4e48-a11f-408ea489df52)
</details>
<details>
<summary>API changelog entry</summary>

![Screenshot from 2023-05-11 19-22-41](https://github.com/zulip/zulip/assets/63245456/0227e676-5917-47cc-b1f2-e0d63c78f7a1)
</details>
<details>
<summary>Example of `failed` field documentation</summary>

#### This description is part of a shared Schema in the OpenAPI documentation, so it will be the same in all other places it appears in the API documentation.

![Screenshot from 2023-05-11 19-22-05](https://github.com/zulip/zulip/assets/63245456/38305e21-f5cf-43e5-a4e0-ae0545db2783)
</details>



---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alexmv commented on zerver/actions/scheduled_messages.py:
```diff
@@ -263,24 +305,36 @@ def try_deliver_one_scheduled_message(logger: logging.Logger) -> bool:
-        scheduled_message.save(update_fields=["failed", "failure_message"])
-        logger.info("Failed with message: %s", scheduled_message.failure_message)
-    except Exception:
-        # An unexpected failure; store as a generic 500 error.
+    except Exception as e:
         scheduled_message.refresh_from_db()
         was_delivered = scheduled_message.delivered
         scheduled_message.failed = True
-        scheduled_message.failure_message = _("Internal server error")
+
+        if isinstance(e, JsonableError):
+            scheduled_message.failure_message = e.msg
+            logger.info("Failed with message: %s", e.msg)
+        else:
+            # An unexpected failure; store and send user a generic
+            # internal server error in notification message.
+            scheduled_message.failure_message = _("Internal server error")
+            logger.exception(
+                "Unexpected error sending scheduled message %s (sent: %s)",
+                scheduled_message.id,
+                was_delivered,
+                stack_info=True,
+            )
```
This needs a `with override_language(user_profile.default_language)` around it, since it's run in a worker, doesn't have a current user, and we want the `_("Internal server error")` as well as the JsonableError possibly raised by `send_scheduled_message` to be in the user's language.

This has the side effect of making the error in our logs also be translated, but I don't think there's a simple way around that, because the JsonableError doesn't store the untranslated string.  We could potentially add that, which would let us defer the translation of this string to display time.

timabbott replied:
Yeah, that sounds correct to me. If we care for our logs, we can add the exception `code` to what we store in the database in the future. And to be clear, the block that needs the override is the full `try` block.

laurynmm replied:
Yeah, my original manual testing for translated strings was flawed. Thanks for pointing this out.

I've updated these changes to that the full try/except block has the override for the sending user's language. Have screenshot of the Notification Bot direct message and log with an existing translated string in the easily triggered "this message send time is too far past" JsonableError in [this comment](https://github.com/zulip/zulip/pull/25553#issuecomment-1546146873).

Should I put the adding the exception `code` into the scheduled message model as a follow up?

timabbott commented on zerver/tests/test_scheduled_messages.py:
```diff
@@ -241,6 +242,14 @@ def test_too_late_to_deliver_scheduled_message(self) -> None:
+            msg = most_recent_message(scheduled_message.sender)
+            self.assertEqual(msg.recipient.type, msg.recipient.PERSONAL)
+            self.assertEqual(msg.sender_id, self.notification_bot(realm).id)
+            self.assertIn("Scheduled send time has already passed", msg.content)
```
Yeah I'm honestly not sure what's best here. I'm 50/50 on whether I think the UI for this will just display a "failed" icon without actually displaying the message.

The actual explanation for this error is of the form "This Zulip Server either got its queue backlogged or otherwise was experiencing a (partial) outage of the send later service when this message was supposed to be sent, and we decided to cancel the message and notify you rather than sending it more than 15 minutes late". I don't know how to explain this to a user concisely.

laurynmm replied:
Updated this for @alya's suggestion [here on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/scheduled.20message.20error.20handling/near/1568030).

timabbott commented on zerver/tests/test_scheduled_messages.py:
```diff
@@ -314,6 +323,14 @@ def test_delivery_type_reminder_failed_to_deliver_scheduled_message_unknown_exce
+            msg = most_recent_message(scheduled_message.sender)
+            self.assertEqual(msg.recipient.type, msg.recipient.PERSONAL)
+            self.assertEqual(msg.sender_id, self.notification_bot(realm).id)
+            self.assertIn("Internal server error", msg.content)
```
I like us to have explicit test coverage the "the message failed because the user was deactivated shouldn't generate a notification bot message".

laurynmm replied:
Added explicit text coverage for both deactivated cases: realm and user.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -17648,6 +17651,13 @@ components:
+          type: boolean
+          description: |
+            Whether the server has tried to send the scheduled message
+            and it failed to successfully send.
```
We can extend this with something about how to use it -- "A client should display these in screens that allow a user to view their scheduled messages and unschedule/copy content from them, both to explain the failure and to allow the user to recovery the message content."

laurynmm replied:
Expanded both this and the API changelog entry. See updated screenshots in [this comment](https://github.com/zulip/zulip/pull/25553#issuecomment-1546146873).

laurynmm commented:
**Updates**:
- Changed the error message text when the scheduled date/time is too far in the past to be: "Message could not be sent at the scheduled time."; see [CZO conversation here](https://chat.zulip.org/#narrow/stream/101-design/topic/scheduled.20message.20error.20handling/near/1568026).
- Fixed the translation issue. The whole try/except block in `try_deliver_one_scheduled_message` has the override for the sending user's language.
  - See updated manual testing screenshot below with an existing translated error string in the JsonableError message field for the above bullet point's situation, which is easy to trigger in the dev environment.
  - This does indeed mean the log will have the translated string, see log from dev environment below.
- Added explicit testing for the deactivated realm and user for not sending the Notification Bot message.
- Expanded the API documentation updates to describe what clients should do with the `failed` field (see screenshots below).
- Added a commit before the API changes that updates the edit scheduled message function to update the `failed` and `failure_message` fields if/when a user edits a scheduled message that the server failed to send. See [this CZO conversation](https://chat.zulip.org/#narrow/stream/378-api-design/topic/scheduled.20messages/near/1567529) for more context on that addition.

**Updated screenshots and console log**:
<details>
<summary>Manual test of translated error string in JsonableError case</summary>

![Screenshot from 2023-05-12 18-44-35](https://github.com/zulip/zulip/assets/63245456/d50b1895-d006-41f3-91b4-3c4799101f57)
</details>
<details>
<summary>Dev environment console log of the above error</summary>

```console
Tornado server (re)started on port 9993
2023-05-12 16:42:44.558 INFO [:9993] Tornado 9993 loaded 1 event queues in 0.000s
2023-05-12 16:42:44.659 INFO [zerver.management.commands.deliver_scheduled_messages] Sending scheduled message 9 with date 2023-05-12 16:36:22+00:00 (sender: 8)
2023-05-12 16:42:44.680 INFO [zerver.management.commands.deliver_scheduled_messages] Failed with message: Ungültiger Datentyp für Empfänger
```
</details>
<details>
<summary>API changelog entry expanded</summary>

![Screenshot from 2023-05-12 19-31-59](https://github.com/zulip/zulip/assets/63245456/3aa6a27d-d2b3-4258-b61e-63e05c4bf082)

</details>
<details>
<summary>API `failed` field description expanded</summary>

![Screenshot from 2023-05-12 20-31-40](https://github.com/zulip/zulip/assets/63245456/3763b53f-3606-4ed6-96b1-d6680c378154)
</details>

timabbott commented:
Made the following tweaks and marked to merge once CI passes, thanks @laurynmm!

```
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index 7ebefb9d30..0afe012f38 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -22,13 +22,13 @@ format used by the Zulip server that they are interacting with.
 
 **Feature level 181**
 
-* [`GET /scheduled_messages`](/api/get-scheduled-messages),
-  [`GET /events`](/api/get-events), [`POST /register`](/api/register-queue):
-  Added `failed` boolean field to scheduled message objects to indicate
-  if the server tried to send the scheduled message and was unsuccessful.
-  Clients that support unscheduling and editing scheduled messages should
-  use this field to indicate to the user when sending a scheduled message
-  has failed, `failed = True`.
+* [`GET /scheduled_messages`](/api/get-scheduled-messages), [`GET
+  /events`](/api/get-events), [`POST /register`](/api/register-queue):
+  Added `failed` boolean field to scheduled message objects to
+  indicate if the server tried to send the scheduled message and was
+  unsuccessful. Clients that support unscheduling and editing
+  scheduled messages should use this field to indicate to the user
+  when a scheduled message failed to send at the appointed time.
 
 **Feature level 180**
 
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index efdc11cadf..cdf30a810e 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -17658,7 +17658,7 @@ components:
             and it failed to successfully send.
 
             Clients that support unscheduling and editing scheduled messages
-            should display scheduled messages with `failed = True` with an
+            should display scheduled messages with `failed = true` with an
             indicator that the server failed to send the message at the
             scheduled time, so that the user is aware of the failure and can
             get the content of the scheduled message.
```

(I felt it unnecessary to spell out `true` in the changelog entry -- it's kinda the obvious meaning of a boolean parameter of that name).
